### ISSUE TITLE
Plugins with library dependencies in PATH fail to load

### ISSUE THREAD

**Issue by snq-** [x64dbg/x64dbg#3047]
snapshot_2023-03-04_02-26 breaks plugins with DLL dependencies in folders specified in the PATH environment variable (eg. x64dbgpy won't load even if python27.dll is available in a folder listed in PATH)

reverting to snapshot_2023-03-03_01-29 fixes the problem

**Comment by mrexodia** [comment]
Hm, yeah this was pretty much the intention of the feature. It changes the way Windows loads DLLs to:

1. First check C:\Windows\System32
2. Then check the application folder

I should be able to disable these checks once plugins start loading though, because at that point everything is user-triggered anyway.

**Comment by mrexodia** [comment]
Thanks, but your workaround completely destroys the whole purpose of the signature checks. If you restore the search path before loading the bridge it’ll re-enable the DLL hijacking that it was fixing.

In general I’m not convinced that DLLs from the PATH should ever be loaded into x64dbg. In the case of x64dbgpy using the full path from the registry is the solution.

**Comment by mrexodia** [comment]
Of course you can easily disable the signature checks by invalidating the signature, this mechanism is only enabled for signed snapshots.

**Comment by mrexodia** [comment]
Probably the best place to do this would be here: https://github.com/x64dbg/x64dbg/blob/7e17c2ccd27ce6088ad906566c998438f5781133/src/dbg/x64dbg.cpp#L765

The 'thread model' (which is fundamentally stupid) is that somebody was using a signed `x32dbg.exe` as a lolbin. They replaced `x32bridge.dll` with some malicious code (since this is loaded when you launch `x32dbg.exe`). The 'fix' is to first load all known dependencies of x64dbg into the process space, while making sure everything is signed and no DLL hijacking is possible. The DLL you need to verify signatures (`wintrust.dll`) is vulnerable to DLL hijacking, so this is why the search order has to be changed.

This makes it so you would have to ship all of x64dbg and add additionally ship `plugins\malware.dp32` to get code execution from a signed executable, which is significantly harder than using `windbg.exe` for instance.

**Comment by mrexodia** [comment]
It looks like this also affects Dotx64dbg. I will need to investigate a bit more if it’s possible to also add the `plugins` directory to the search list…

**Comment by mrexodia** [comment]
Yeah there is also a plan to allow a separate folder for each plugin. Loading something from the user PATH is always a bad idea though, that should be loaded using an absolute path instead.

**Comment by torusrxxx** [comment]
It seems Python-based tools generally expect either Python itself is shipped, or it is available at user PATH. I would like to load a DLL from PATH by my private plugin too because it could also be in use by another tool, and I don't like copying it everywhere.

**Comment by mrexodia** [comment]
You can load it using an absolute path from the PATH. The issue with loading random dependencies from the user’s PATH is that you can easily hijkack them

**Comment by mrexodia** [comment]
I added the `plugins` folder to the DLL search path, which should solve the majority of the issues. Will fix x64dbgpy next. I will review the restoration of the original search order later.
